helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java

```diff
@@ -0,0 +1,181 @@
+      _description = description;
+      _pattern = Arrays.asList(patterns);
+      _warningLimit = warningLimit;
+      _limit = limit;
+    }
+  }
+
+  public static boolean afterClassCheck(String classname) {
+    ZkTestBase.reportPhysicalMemory();
+    // step 1: get all active threads
+    List<Thread> threads = getAllThreads();
+    System.out.println(classname + " has active threads cnt:" + threads.size());
+
+    // step 2: categorize threads
+    Map<String, List<Thread>> threadByName = null;
+    Map<ThreadCategory, Integer> threadByCnt = new HashMap<>();
+    Map<ThreadCategory, Set<Thread>> threadByCat = new HashMap<>();
+    try {
+      threadByName = threads.
+          stream().
+          filter(p -> p.getThreadGroup() != null && p.getThreadGroup().getName() != null
+              &&  ! "system".equals(p.getThreadGroup().getName())).
+          collect(Collectors.groupingBy(p -> p.getName()));
+    } catch (Exception e) {
+      System.out.println("filtering thread failure with exception:" + e.getStackTrace());
+    }
+
+    threadByName.entrySet().stream().forEach(entry -> {
+      String key = entry.getKey(); // thread name
+      Arrays.asList(ThreadCategory.values()).stream().forEach(category -> {
+        if (category.getMatchPred().test(key)) {
+          Integer count = threadByCnt.containsKey(category) ? threadByCnt.get(category) : 0;
+          threadByCnt.put(category, count + entry.getValue().size());
+          Set<Thread> thisSet = threadByCat.getOrDefault(category, new HashSet<>());
+          thisSet.addAll(entry.getValue());
+          threadByCat.put(category, thisSet);
+        }
+      });
+    });
+
+    // step 3: enforce checking policy
+    boolean checkStatus = true;
+    for (ThreadCategory threadCategory : ThreadCategory.values()) {
+      int limit = threadCategory.getLimit();
+      int warningLimit = threadCategory.getWarningLimit();
+
+      Integer catThreadCnt = threadByCnt.get(threadCategory);
+      if (catThreadCnt != null) {
+        boolean dumpThread = false;
+        if (catThreadCnt > limit) {
+          checkStatus = false;
+          System.out.println(
+              "Failure " + threadCategory.getDescription() + " has " + catThreadCnt + " thread");
+          dumpThread = true;
+        } else if (catThreadCnt > warningLimit) {
+          System.out.println(
+              "Warning " + threadCategory.getDescription() + " has " + catThreadCnt + " thread");
+          dumpThread = true;
+        } else {
+          System.out.println(threadCategory.getDescription() + " has " + catThreadCnt + " thread");
+        }
+        if (!dumpThread) {
+          continue;
+        }
+        // print first 10 thread names
+        int i = 0;
+        for (Thread t : threadByCat.get(threadCategory)) {
+          System.out.println(i + " thread:" + t.getName());
+          i++;
+          if (i == 100) {
```

Comment:
Comments said "10", but I think "100" makes more sense.

Reply:
changed the comments